### ISSUE TITLE
Order of union members in SchemaFor is unstable for sealed traits

### ISSUE THREAD

**Issue by alexjg** [sksamuel/avro4s#343]
I'm encountering an issue where the order of union members in the `SchemaFor` is unstable. Unfortunately I'm struggling to put together a minimal failing example.

As I understand it the order of union members comes from this code in `SchemaFor.scala`:

```scala
  def dispatch[T: WeakTypeTag](ctx: SealedTrait[Typeclass, T]): SchemaFor[T] = new SchemaFor[T] {
    override def schema(namingStrategy: NamingStrategy) = {

      import scala.reflect.runtime.universe

      val runtimeMirror = universe.runtimeMirror(getClass.getClassLoader)
      val tpe = runtimeMirror.weakTypeOf[T]
      val objs = tpe.typeSymbol.isClass && tpe.typeSymbol.asClass.knownDirectSubclasses.forall(_.isModuleClass)

      if (objs) {
        val symbols = ctx.subtypes.map { sub =>
          val namer = Namer(sub.typeName, sub.annotations)
          namer.name
        }
        val namer = Namer(ctx.typeName, ctx.annotations)
        SchemaBuilder.enumeration(namer.name).namespace(namer.namespace).symbols(symbols: _*)
      } else {
        val schemas = ctx.subtypes.map(_.typeclass.schema(namingStrategy))
        SchemaHelper.createSafeUnion(schemas: _*)
      }
    }
  }
```

And `SealedTrait.subtypes` is implemented in terms of `knownDirectSubclasses`, which is notoriously unreliable. Does anyone know of any tips to make it more consistent?

**Comment by sksamuel** [comment]
Which version ?

**Comment by alexjg** [comment]
I'm actually running from my own fork (which is up to date with `master`) because I need a fix relating to enum name generation (https://github.com/sksamuel/avro4s/pull/332).

I'll try running with the latest snapshot now and report back.

**Comment by alexjg** [comment]
Yep, still experiencing the issue on the latest snapshot. The order changes depending on which files have been most recently changed, so I suspect it's a dependency on compilation order. I've heard that one way to fix this is to move the files containing the sealed traits into a separate project so that they are definitely compiled before the files containing the macro invocations.

Unfortunately I can't find much information by googleing about compilation order so it's all a bit of trial and error. I raised this issue in case it was something anyone else was familiar with.

**Comment by alexjg** [comment]
Splitting the files I was generating schemas from into a separate project fixed the issue. This seems like it's more an issue with the way Magnolia works than specific to Avro4s so I'm going to close this.

**Comment by alexjg** [comment]
I'm reopening this. I've run into the issue again. It seems that splitting into several projects didn't actually fix the issue, it just accidentally solved the issue by changing something in the compilation order (I think).

The root of the problem is that we're relying on `knownDirectSubclasses` to return a stable ordering for subclasses of a sealed trait. Unfortunately it returns a `Set`, which means that we shouldn't rely on it. The fact that it's mostly worked so far is an accident.

The first thing that occurs to me is that we could sort the subtypes alphabetically. This would give us a deterministic ordering but it would make it impossible to use with `avro4s` with schemas which are ordered arbitrarily. Maybe we could add an annotation for the position in a union or enum, something like `AvroSortPriority(priority: float)` might do the trick?

**Comment by obogobo** [comment]
I think we might be encountering this as well, small test harness:

` 
def serialize(event: NormalizedEventEnvelope): Array[Byte] = {
    val baos = new ByteArrayOutputStream()
    val output = AvroOutputStream
      .binary[NormalizedEventEnvelope]
      .to(baos)
      .build(AvroSchema[NormalizedEventEnvelope])

    output.write(event)
    output.close()

    baos.toByteArray
  }

  def deserialize(bytes: Array[Byte]): NormalizedEventEnvelope = {
    val input = AvroInputStream
      .binary[NormalizedEventEnvelope]
      .from(bytes)
      .build(AvroSchema[NormalizedEventEnvelope])
    
    input.iterator.next()
  }

case class NormalizedEventEnvelope(
  ...somestuff
  event: NormalizedEvent # sealed trait hierarchy
)
`

the above fails with an exception: "Cannot find subschema for ..." 
...but weirdly, runs just fine if the `AvroSchema[NormalizedEventEnvelope]` is hoisted out of the .build() in serialize/deserialize and brought to a val in scope for both functions. 

If they both reference the same *instance* of that schema there are no problems - which seems potentially related to an ordering issue with UNION types?

**Comment by alexjg** [comment]
That seems plausible, you could try using the `AvroSortPriority` annotation in the PR I've linked to this issue. If that solves the problem that would be pretty good evidence that you're running into this issue.

**Comment by alexjg** [comment]
This is fixed by #357 so I'm closing it.
